packages/cyberstorm/src/components/MultiSelectSearch/MultiSelectSearch.tsx

```diff
@@ -0,0 +1,191 @@
+"use client";
+import React from "react";
+import styles from "./MultiSelectSearch.module.css";
+import { classnames } from "../../utils/utils";
+import { Button, Icon } from "../../index";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import {
+  faCaretDown,
+  faCircleXmark,
+  faXmark,
+} from "@fortawesome/pro-solid-svg-icons";
+import { isNode } from "../../utils/type_guards";
+
+export type MultiSelectSearchOption = {
+  label: string;
+  value: string;
+};
+type Props = {
+  options: MultiSelectSearchOption[];
+  value: MultiSelectSearchOption[];
+  onChange: (v: MultiSelectSearchOption[]) => void;
+  onBlur: () => void;
+  // TODO: Implement disabled state
+  disabled?: boolean;
+  placeholder?: string;
+  color?: "red" | "green";
+};
+
+/**
+ * Cyberstorm MultiSelectSearch component
+ *  TODO: Better keyboard navigation
+ */
+export const MultiSelectSearch = React.forwardRef<HTMLInputElement, Props>(
+  function MultiSelectSearch(props, ref) {
+    const { options, value, onChange, onBlur, placeholder, color } = props;
+    const menuRef = React.useRef<HTMLDivElement | null>(null);
+    const inputRef = React.useRef<HTMLInputElement | null>(null);
+
+    const [isVisible, setIsVisible] = React.useState(false);
+    const [search, setSearch] = React.useState("");
+    const [filteredOptions, setFilteredOptions] = React.useState(options);
+
+    function add(incomingOption: MultiSelectSearchOption) {
+      if (!value.some((option) => option.value === incomingOption.value)) {
+        onChange([...value, incomingOption]);
+      }
+    }
+
+    function remove(incomingOption: MultiSelectSearchOption) {
+      onChange(value.filter((option) => option.value !== incomingOption.value));
+    }
+
+    const hideMenu = React.useCallback(
+      (e: MouseEvent | TouchEvent) => {
+        if (
+          menuRef.current &&
+          isVisible &&
+          !menuRef.current.contains(isNode(e.target) ? e.target : null)
+        ) {
+          setIsVisible(false);
+          onBlur();
+        }
+      },
+      [setIsVisible, isVisible, onBlur, menuRef]
+    );
+
+    // Event listeners for closing menu when clicking or touching outside.
+    React.useEffect(() => {
+      document.addEventListener("mousedown", hideMenu);
+      document.addEventListener("touchstart", hideMenu);
+      return () => {
+        document.removeEventListener("mousedown", hideMenu);
+        document.removeEventListener("touchstart", hideMenu);
+      };
+    });
+
+    React.useEffect(() => {
+      const updatedOptions = options.filter(
+        (option) =>
+          !value.includes(option) &&
+          option.label.toLowerCase().includes(search.toLowerCase())
+      );
+      setFilteredOptions(updatedOptions);
+    }, [value, search, options, setFilteredOptions]);
+
+    return (
+      <div className={styles.root} ref={ref}>
+        <div className={styles.selected}>
+          {value.map((option) => {
```

Comment:
Currently the options are listed in the order they are selected. IDK if it would be more user friendly if they were ordered alphabetically? Something to think about e.g. when fixing the other TODOs later on?

Reply:
Added sorting 👍